src/Java.Interop.Tools.JavaTypeSystem/JavaModels/JavaClassModel.cs

```diff
@@ -0,0 +1,118 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Java.Interop.Tools.JavaTypeSystem.Models
+{
+	public class JavaClassModel : JavaTypeModel
+	{
+		IDictionary<JavaTypeReference, JavaTypeReference>? generic_inheritance_mapping;
+
+		public string BaseType { get; }
+		public string BaseTypeGeneric { get; }
+		public string BaseTypeJni { get; }
+
+		public JavaTypeReference? BaseTypeReference { get; private set; }
+		public List<JavaConstructorModel> Constructors { get; } = new List<JavaConstructorModel> ();
+
+		public JavaClassModel (JavaPackage javaPackage, string javaNestedName, string javaVisibility, bool javaAbstract, bool javaFinal, string javaBaseType, string javaBaseTypeGeneric, string javaDeprecated, bool javaStatic, string jniSignature, string baseTypeJni) :
+			base (javaPackage, javaNestedName, javaVisibility, javaAbstract, javaFinal, javaDeprecated, javaStatic, jniSignature)
+		{
+			BaseType = javaBaseType;
+			BaseTypeGeneric = javaBaseTypeGeneric;
+			BaseTypeJni = baseTypeJni;
+		}
+
+		public override void Resolve (JavaTypeCollection types, List<JavaUnresolvableModel> unresolvables)
+		{
+			var type_parameters = GetApplicableTypeParameters ().ToArray ();
+
+			// Resolve base class
+			if (FullName != "java.lang.Object" && FullName != "java.lang.Throwable") {
```

Comment:
Based on this context -- still not even half way through reviewing! -- I think there shouldn't be a `FullName` property, as (1) without more context, it's uncertain what *kind* of "full name" this is, and (2) in *this* context, I have to assume that `FullName` is either a Java name, e.g. `java.lang.Thread.State`, or some "unholy" intermixing of Java `.`-separated package names and `$`-separated inner names, e.g. `java.lang.Thread$State`.

We *need* a way of knowing, unambiguously, if a type is nested or not, and what its package is.  JNI names do that.  Java full names *don't*.  An "unholy" mix allows knowing about nested types, but is otherwise "new" and separate from other pre-existing constructs.

I would thus prefer renaming the `FullName` property to `JniFullName`, and using JNI semantics consistently.  (And/or rename `FullName` to `JavaFullName`, so that semantics are clearer.)

Reply:
I agree that all these different types of names can be extremely confusing.  Ultimately I don't think naming is going to ever be enough to keep them straight.  I went with providing `<summary>` comments on them to be very explicit about what each one means:

https://github.com/xamarin/java.interop/pull/849/files#diff-9805a3d07252c317909b00c28c6f05fb1e0ecbc9355e99f88e31fc23d3fa05e1R54-R57